plutus-tx/src/PlutusTx/Plugin/IsString.hs

```diff
@@ -0,0 +1,40 @@
+{-# LANGUAGE KindSignatures #-}
+
+module PlutusTx.Plugin.IsString (
+  Utf8Encoded (..),
+  RawBytes (..),
+  LiteralEncoding (..),
+) where
+
+import Prelude
+
+import Data.Kind (Type)
+import Data.String (IsString (..))
+
+{- Note [IsString instances and UTF-8 encoded string literals]
+
+GHC Core encodes string literals in UTF-8 by default.
+Plugin handles GHC Core expressions representing such literals specially:
+in compile type it undoes the UTF-8 encoding when string literal is used to construct
+a value of type 'BuiltinByteString' and preserves the UTF-8 encoding to construct 'BuiltinString'.
+
+The 'LiteralEncoding' type helps to distinguish between these two cases.
+
+The newtypes 'Utf8Encoded' and 'RawBytes' are used to indirectly provide IsString instances
+for other newtypes that wrap 'BuiltinString' or 'BuiltinByteString' i.e. 'TokenName':
+  - '"Мой Токен" :: TokenName'
+  - 'runRawBytes "Мой Токен" :: TokenName'
+
+-}
+
+data LiteralEncoding = UndoUtf8 | PreserveUtf8
```

Comment:
The idea here is to model 2 possible behaviours wrt. UTF8 encoding done by the GHC: 
- Undo it.
- Preserve it.

Maybe I should rename the type itself, say `EncodingAction` or `GhcEncodingPreservation`

Reply:
Lets take a syntax like "\xBA\xBE" (or "\186\190") that represents two bytes BA  (186), BE (190)

When I say "raw" I mean that there is 1:1 correspondence between two bytes in string and two bytes in resulting ByteString.

In order to get this 1:1 correspondence we need to undo the UTF8 encoding that is already done in GHC core:
```haskell
unpackCStringUtf8# "\194\186\194\190"#
```

This isn't 1:1 out of the box: the byte BA is represented as two bytes "\194\186" and byte BE is represented as "\194\190"

By "un-doing" UTF8 encoding I mean converting "\194\186" to "\186" and "\194\190" to "\190" to restore the 1:1 correspondence.

Reply:
`\186` is not a byte, it is a char (or unicode code point). And it can go beyond 255: `\12345` is also a char.

To convert this char to byte(s) requires an encoding. How about call it "Wrapping" or something like that, to indicate that it wraps around if it is >=256 (which is what ByteString does)?

Reply:
I am intentionally trying to get away from this wrapping around behaviour (which I believe hurts UX) in favor of the fail-fast behaviour. I don't think that being consistent with the GHC in its quirks is better than having a less surprising behaviour for users.